### ISSUE TITLE
Get-PartnerCustomerOffer - Returns LIST instead of Array

### ISSUE THREAD

**Issue by cblackuk** [Microsoft/Partner-Center-PowerShell#22]
We are trying to convert from the old module and the command we used before was:

```powershell
Get-PCOffer -CountryId "GB" -LocaleId "en-US"
```

This returns PowerShell objects like a normal cmdlet so you can pass Where-Object to it.

It was achieved by this code "transformation"

```powershell
function _applyTypes {
   param($item,$type)
   $item.PSObject.TypeNames.Insert(0, $type)
}

function _formatResult {
    param ($obj, $type)
    try 
    {
        if($obj.items)
        {
            foreach($item in $obj.items){_applyTypes -item $item -type $type}
            return $obj.items
        }
    }
    catch
    {
        _applyTypes -item $obj -type $type
        return $obj
    }
}
```

Inside the command:

```powershell
            $response = Invoke-RestMethod -Uri $url -Headers $headers -ContentType "application/json" -Method "GET" #-Debug -Verbose
            $obj += $response.Substring(1) | ConvertFrom-Json
            return (_formatResult -obj $obj -type "Offer")
```

New command however returns a LIST... so you cannot treat is a an PS Object.

```powershell
Get-PartnerOffer -CountryCode "GB"  | Where-Object {$_.Name -like "Microsoft Azure"} 
```
If you run this you will get everything returned.

Example:
```powershell
$OfferTest = Get-PartnerOffer -CountryCode "GB"  | Where-Object {$_.Name -like "Microsoft Azure"}
PS C:\> $OfferTest = Get-PartnerOffer -CountryCode "GB"  | Where-Object {$_.Name -like "Microsoft Azure"}
$OfferTest.GetType()

IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     True     List`1                                   System.Object
```

This is because the format is LIST in CSharp code here:
https://github.com/Microsoft/Partner-Center-PowerShell/blob/4b03067143bd80204e11394cd04c40a2c015405e/src/PowerShell/Commands/GetPartnerOffer.cs

```C#
            WriteObject(
                offers
                    .Where(o => o.IsAddOn == isAddOn && o.IsTrial == isTrial)
                    .Select(o => new PSOffer(o))
                    .ToList());
```

Workaround is to pass it through a loop:

```powershell

$ArrayTest = @()
 ForEach ($OfferStuff in $OfferTest)
 {
    $ArrayTest += $OfferStuff

 }
 $ArrayTest |  Where-Object {$_.Name -like "Microsoft Azure"}


OfferId                : MS-AZR-0145P
Name                   : Microsoft Azure
Billing                : Usage
MaximumQuantity        : 999999999
SupportedBillingCycles : {Monthly}
IsAddOn                : False
IsTrial                : False
```

Can the object be changed in CSharp from a list to an array please so it behaves like proper PowerShell cmdlet?

**Comment by cblackuk** [comment]
I found a better workaround:
```powershell
$Offer = Get-PartnerOffer -CountryCode "GB" 
$Offer = $Offer | Where-Object {$_.Name -like "Microsoft Azure"}
$Offer
```
Output:

```powershell

OfferId                : MS-AZR-0145P
Name                   : Microsoft Azure
Billing                : Usage
MaximumQuantity        : 999999999
SupportedBillingCycles : {Monthly}
IsAddOn                : False
IsTrial                : False
```

So it only breaks when you try to filter at the same time as you are trying to list offers but once the object has been saved, it can be treated as PS Object.

Still needs fixing in CSharp imho.

**Comment by isaiahwilliams** [comment]
Hi @cblackuk, 

Thank you for bring this to my attention. The reason you are encountering this behavior is that PowerShell is writing the collection instead of each individual line. I have corrected this issue by modifing the WriteOutput code to look like 

```csharp
        private void WriteOutput(IEnumerable<Offer> offers)
        {
            bool isAddOn = AddOn.IsPresent && AddOn.ToBool();
            bool isTrial = Trial.IsPresent && Trial.ToBool();

            WriteObject(
                offers
                    .Where(o => o.IsAddOn == isAddOn && o.IsTrial == isTrial)
                    .Select(o => new PSOffer(o)), true);
        }
```

This issue will be resolved in the next release of the module. My hope is to have it ready to go this week. In the meantime please let me know if you discover any other issues.

**Comment by cblackuk** [comment]
That is AMAZING!!! Thank you ever so much for getting back to me!!!

The second issue I raised is much bigger one and a showstopper for me ;-( @isaiahwilliams

**Comment by isaiahwilliams** [comment]
Hi @cblackuk, 

I just published version [1.5.1810.2](https://www.powershellgallery.com/packages/PartnerCenter/1.5.1810.2) which includes a fix for this issue. Once you update to the new version, you will be able to filter the items as you except. Please let me know if you have any issues.
